core/src/main/java/hudson/slaves/NodeProvisioner.java

```diff
@@ -217,29 +218,43 @@ public void run() {
-                                Node node = f.future.get();
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onComplete(f, node);
-                                }
-    
-                                jenkins.addNode(node);
-                                LOGGER.log(Level.INFO,
-                                        "{0} provisioning successfully completed. " 
-                                                + "We have now {1,number,integer} computer(s)",
-                                        new Object[]{f.displayName, jenkins.getComputers().length});
-                            } catch (InterruptedException e) {
-                                throw new AssertionError(e); // since we confirmed that the future is already done
-                            } catch (ExecutionException e) {
-                                LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch",
-                                        e.getCause());
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e.getCause());
+                                Node node = null;
+                                try {
+                                    node = f.future.get();
+                                } catch (InterruptedException e) {
+                                    throw new AssertionError("InterruptedException occurred", e); // since we confirmed that the future is already done
+                                } catch (ExecutionException e) {
+                                    Throwable cause = e.getCause();
+                                    if(!(cause instanceof AbortException)) {
+                                        LOGGER.log(Level.WARNING, "Unexpected exception encountered while "
+                                                + "provisioning agent " + f.displayName, cause);
+                                    }
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onFailure(f, cause);
+                                    }
+                                    node = null;
                                 }
-                            } catch (IOException e) {
-                                LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch", 
-                                        e);
-                                for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
-                                    cl.onFailure(f, e);
+
+                                if (node != null) {
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.fireOnComplete(f, node);
```

Comment:
This probably should have been `fireOnProvisioned` then `fireOnComplete` can take place when the node has been saved... the issue as I see it is if plugins are currently using `onComplete` to modify the node before save... given that https://github.com/search?q=org%3Ajenkinsci+CloudProvisioningListener&type=Code doesn't indicate that there are many people using this extension point... I have also analyzed the CloudBees internal usage and none of the 3 closed source implementations we have would object to the splitting.

Thus I suggest that we can change the listener contract and move onCompleted to after the save and introduce onProvisioned here